{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15013 2456\n"
     ]
    }
   ],
   "source": [
    "# Import data sources - HPA - DAVID - ORA - DisGeNet\n",
    "hpa_data = pd.read_csv('../Data/HPA/Raw/NOT.tsv', sep='\\t').drop_duplicates(subset='Gene')\n",
    "dgn_data = pd.read_csv('../Data/NIH/Unique Values Per Positive Label.csv')\n",
    "dgn_labels = pd.read_csv('../Data/NIH/DisGeNet_Labels_ensembl_conversion.csv')\n",
    "dgn_labels.columns = ['Gene','Ensembl','Score_gda']\n",
    "print(len(hpa_data.index), len(dgn_data.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Feature Sparsity:\n",
      " Gene                                                     0\n",
      "Ensembl                                                  0\n",
      "Protein class                                            0\n",
      "Biological process                                    6858\n",
      "Molecular function                                    6651\n",
      "Disease involvement                                  10548\n",
      "Subcellular location                                  3666\n",
      "Tissue RNA - lung [NX]                                   0\n",
      "Single Cell Type RNA - Mucus-secreting cells [NX]        0\n",
      "dtype: int64\n"
     ]
    }
   ],
   "source": [
    "identifiers = [\n",
    "    \"Gene\",\n",
    "    \"Ensembl\"\n",
    "]\n",
    "discrete_features = [\n",
    "    \"Protein class\",\n",
    "    \"Biological process\",\n",
    "    \"Molecular function\",\n",
    "    \"Disease involvement\", # consider gettting rid of this feature\n",
    "    \"Subcellular location\",\n",
    "]\n",
    "continuous_features = [\n",
    "    \"Tissue RNA - lung [NX]\",\n",
    "    \"Single Cell Type RNA - Mucus-secreting cells [NX]\"\n",
    "]\n",
    "\n",
    "# Create new DF filtering for features of interest\n",
    "hpa_features = hpa_data.loc[:, hpa_data.columns.isin(identifiers+discrete_features+continuous_features)]\n",
    "\n",
    "# Check NAs for each feature\n",
    "print(\"Feature Sparsity:\\n\", hpa_features.isna().sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "           Gene          Ensembl  Score_gda\n",
      "3732      DPCR1  ENSG00000168631        0.1\n",
      "5528       H1FX  ENSG00000184897        0.2\n",
      "5722  HIST1H2BE  ENSG00000274290        0.3\n",
      "7891     MT-CO2  ENSG00000198712        0.1\n"
     ]
    }
   ],
   "source": [
    "# leftover from comparing label sets\n",
    "t2 = pd.merge(hpa_data, dgn_labels[['Gene','Score_gda']], on='Gene', how='left')\n",
    "t3 = pd.merge(hpa_data, dgn_labels[['Ensembl','Score_gda']], on='Ensembl', how='left')\n",
    "t2_m = t2.loc[~t2.Score_gda.isna()].drop_duplicates(subset='Gene')[['Gene','Ensembl','Score_gda']]\n",
    "t3_m = t3.loc[~t3.Score_gda.isna()].drop_duplicates(subset='Gene')[['Gene','Ensembl','Score_gda']]\n",
    "extra = t3_m.loc[~t3_m.Ensembl.isin(t2_m.Ensembl)]\n",
    "print(extra)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1826\n",
      "1830\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Ensembl</th>\n",
       "      <th>Protein class</th>\n",
       "      <th>Biological process</th>\n",
       "      <th>Molecular function</th>\n",
       "      <th>Disease involvement</th>\n",
       "      <th>Subcellular location</th>\n",
       "      <th>Tissue RNA - lung [NX]</th>\n",
       "      <th>Single Cell Type RNA - Mucus-secreting cells [NX]</th>\n",
       "      <th>Score_gda</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A2M</td>\n",
       "      <td>ENSG00000175899</td>\n",
       "      <td>Cancer-related genes, Candidate cardiovascular...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Protease inhibitor, Serine protease inhibitor</td>\n",
       "      <td>Cancer-related genes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>227.4</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>ENSG00000128274</td>\n",
       "      <td>Enzymes, Predicted membrane proteins</td>\n",
       "      <td>Lipid biosynthesis, Lipid metabolism</td>\n",
       "      <td>Glycosyltransferase, Transferase</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mitochondria</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAAS</td>\n",
       "      <td>ENSG00000094914</td>\n",
       "      <td>Disease related genes, Potential drug targets,...</td>\n",
       "      <td>mRNA transport, Protein transport, Translocati...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Disease mutation</td>\n",
       "      <td>Nuclear membrane,Centrosome,Cytosol</td>\n",
       "      <td>13.4</td>\n",
       "      <td>11.5</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AACS</td>\n",
       "      <td>ENSG00000081760</td>\n",
       "      <td>Enzymes, Predicted membrane proteins</td>\n",
       "      <td>Fatty acid metabolism, Lipid metabolism</td>\n",
       "      <td>Ligase</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Vesicles</td>\n",
       "      <td>4.9</td>\n",
       "      <td>20.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AADAC</td>\n",
       "      <td>ENSG00000114771</td>\n",
       "      <td>Enzymes, Predicted intracellular proteins</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hydrolase</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15008</th>\n",
       "      <td>ZXDB</td>\n",
       "      <td>ENSG00000198455</td>\n",
       "      <td>Predicted intracellular proteins, Transcriptio...</td>\n",
       "      <td>Transcription, Transcription regulation</td>\n",
       "      <td>Activator</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Nucleoplasm</td>\n",
       "      <td>2.3</td>\n",
       "      <td>6.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15009</th>\n",
       "      <td>ZXDC</td>\n",
       "      <td>ENSG00000070476</td>\n",
       "      <td>Predicted intracellular proteins, Transcriptio...</td>\n",
       "      <td>Transcription, Transcription regulation</td>\n",
       "      <td>Activator</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Nucleoli</td>\n",
       "      <td>10.6</td>\n",
       "      <td>5.9</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15010</th>\n",
       "      <td>ZYG11B</td>\n",
       "      <td>ENSG00000162378</td>\n",
       "      <td>Predicted intracellular proteins</td>\n",
       "      <td>Ubl conjugation pathway</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Golgi apparatus,Intermediate filaments</td>\n",
       "      <td>7.7</td>\n",
       "      <td>10.9</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15011</th>\n",
       "      <td>ZYX</td>\n",
       "      <td>ENSG00000159840</td>\n",
       "      <td>Plasma proteins, Predicted intracellular proteins</td>\n",
       "      <td>Cell adhesion, Host-virus interaction</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Plasma membrane,Actin filaments,Focal adhesion...</td>\n",
       "      <td>52.8</td>\n",
       "      <td>19.9</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15012</th>\n",
       "      <td>ZZEF1</td>\n",
       "      <td>ENSG00000074755</td>\n",
       "      <td>Predicted membrane proteins</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Nucleoplasm,Mitochondria</td>\n",
       "      <td>12.1</td>\n",
       "      <td>30.4</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15013 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Gene          Ensembl  \\\n",
       "0         A2M  ENSG00000175899   \n",
       "1      A4GALT  ENSG00000128274   \n",
       "2        AAAS  ENSG00000094914   \n",
       "3        AACS  ENSG00000081760   \n",
       "4       AADAC  ENSG00000114771   \n",
       "...       ...              ...   \n",
       "15008    ZXDB  ENSG00000198455   \n",
       "15009    ZXDC  ENSG00000070476   \n",
       "15010  ZYG11B  ENSG00000162378   \n",
       "15011     ZYX  ENSG00000159840   \n",
       "15012   ZZEF1  ENSG00000074755   \n",
       "\n",
       "                                           Protein class  \\\n",
       "0      Cancer-related genes, Candidate cardiovascular...   \n",
       "1                   Enzymes, Predicted membrane proteins   \n",
       "2      Disease related genes, Potential drug targets,...   \n",
       "3                   Enzymes, Predicted membrane proteins   \n",
       "4              Enzymes, Predicted intracellular proteins   \n",
       "...                                                  ...   \n",
       "15008  Predicted intracellular proteins, Transcriptio...   \n",
       "15009  Predicted intracellular proteins, Transcriptio...   \n",
       "15010                   Predicted intracellular proteins   \n",
       "15011  Plasma proteins, Predicted intracellular proteins   \n",
       "15012                        Predicted membrane proteins   \n",
       "\n",
       "                                      Biological process  \\\n",
       "0                                                    NaN   \n",
       "1                   Lipid biosynthesis, Lipid metabolism   \n",
       "2      mRNA transport, Protein transport, Translocati...   \n",
       "3                Fatty acid metabolism, Lipid metabolism   \n",
       "4                                                    NaN   \n",
       "...                                                  ...   \n",
       "15008            Transcription, Transcription regulation   \n",
       "15009            Transcription, Transcription regulation   \n",
       "15010                            Ubl conjugation pathway   \n",
       "15011              Cell adhesion, Host-virus interaction   \n",
       "15012                                                NaN   \n",
       "\n",
       "                                  Molecular function   Disease involvement  \\\n",
       "0      Protease inhibitor, Serine protease inhibitor  Cancer-related genes   \n",
       "1                   Glycosyltransferase, Transferase                   NaN   \n",
       "2                                                NaN      Disease mutation   \n",
       "3                                             Ligase                   NaN   \n",
       "4                                          Hydrolase                   NaN   \n",
       "...                                              ...                   ...   \n",
       "15008                                      Activator                   NaN   \n",
       "15009                                      Activator                   NaN   \n",
       "15010                                            NaN                   NaN   \n",
       "15011                                            NaN                   NaN   \n",
       "15012                                            NaN                   NaN   \n",
       "\n",
       "                                    Subcellular location  \\\n",
       "0                                                    NaN   \n",
       "1                                           Mitochondria   \n",
       "2                    Nuclear membrane,Centrosome,Cytosol   \n",
       "3                                               Vesicles   \n",
       "4                                                    NaN   \n",
       "...                                                  ...   \n",
       "15008                                        Nucleoplasm   \n",
       "15009                                           Nucleoli   \n",
       "15010             Golgi apparatus,Intermediate filaments   \n",
       "15011  Plasma membrane,Actin filaments,Focal adhesion...   \n",
       "15012                           Nucleoplasm,Mitochondria   \n",
       "\n",
       "       Tissue RNA - lung [NX]  \\\n",
       "0                       227.4   \n",
       "1                        14.0   \n",
       "2                        13.4   \n",
       "3                         4.9   \n",
       "4                         1.7   \n",
       "...                       ...   \n",
       "15008                     2.3   \n",
       "15009                    10.6   \n",
       "15010                     7.7   \n",
       "15011                    52.8   \n",
       "15012                    12.1   \n",
       "\n",
       "       Single Cell Type RNA - Mucus-secreting cells [NX]  Score_gda  \n",
       "0                                                    0.0        NaN  \n",
       "1                                                    0.0        NaN  \n",
       "2                                                   11.5        NaN  \n",
       "3                                                   20.1        NaN  \n",
       "4                                                    0.0        NaN  \n",
       "...                                                  ...        ...  \n",
       "15008                                                6.1        NaN  \n",
       "15009                                                5.9        NaN  \n",
       "15010                                               10.9        NaN  \n",
       "15011                                               19.9        NaN  \n",
       "15012                                               30.4        NaN  \n",
       "\n",
       "[15013 rows x 10 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hpa_label_data = pd.merge(hpa_features, dgn_data[['Gene','Score_gda']], on='Gene', how='left')\n",
    "print(len(hpa_label_data.loc[~hpa_label_data.Score_gda.isna()].index))\n",
    "\n",
    "for index, row in extra.iterrows() :\n",
    "    hpa_label_data.at[index, 'Score_gda'] = row.Score_gda\n",
    "print(len(hpa_label_data.loc[~hpa_label_data.Score_gda.isna()].index))\n",
    "\n",
    "hpa_label_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    14613\n",
       "1      400\n",
       "Name: Score_gda, dtype: int64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# fill in na's in labels column and turn to 01\n",
    "hpa_label_data.Score_gda = hpa_label_data.Score_gda.fillna(0)\n",
    "hpa_label_data.Score_gda = hpa_label_data.Score_gda.apply(lambda x: 1 if x >= 0.05 else 0)\n",
    "hpa_label_data.Score_gda.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Standardize expression data\n",
    "hpa_label_data[\"Tissue RNA - lung [NX]\"] = (hpa_label_data[\"Tissue RNA - lung [NX]\"] - hpa_label_data[\"Tissue RNA - lung [NX]\"].mean()) / hpa_label_data[\"Tissue RNA - lung [NX]\"].std()\n",
    "col = \"Single Cell Type RNA - Mucus-secreting cells [NX]\"\n",
    "hpa_label_data[col] = (hpa_label_data[col] - hpa_label_data[col].mean()) / hpa_label_data[col].std()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Ensembl</th>\n",
       "      <th>Protein class</th>\n",
       "      <th>Biological process</th>\n",
       "      <th>Molecular function</th>\n",
       "      <th>Disease involvement</th>\n",
       "      <th>Subcellular location</th>\n",
       "      <th>Tissue RNA - lung [NX]</th>\n",
       "      <th>Single Cell Type RNA - Mucus-secreting cells [NX]</th>\n",
       "      <th>Score_gda</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A2M</td>\n",
       "      <td>ENSG00000175899</td>\n",
       "      <td>[Cancer-relatedgenes, Candidatecardiovasculard...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[Proteaseinhibitor, Serineproteaseinhibitor]</td>\n",
       "      <td>[Cancer-relatedgenes]</td>\n",
       "      <td>[]</td>\n",
       "      <td>11.006867</td>\n",
       "      <td>-0.105024</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>ENSG00000128274</td>\n",
       "      <td>[Enzymes, Predictedmembraneproteins]</td>\n",
       "      <td>[Lipidbiosynthesis, Lipidmetabolism]</td>\n",
       "      <td>[Glycosyltransferase, Transferase]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[Mitochondria]</td>\n",
       "      <td>-0.000111</td>\n",
       "      <td>-0.105024</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAAS</td>\n",
       "      <td>ENSG00000094914</td>\n",
       "      <td>[Diseaserelatedgenes, Potentialdrugtargets, Pr...</td>\n",
       "      <td>[mRNAtransport, Proteintransport, Translocatio...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[Diseasemutation]</td>\n",
       "      <td>[Nuclearmembrane, Centrosome, Cytosol]</td>\n",
       "      <td>-0.031058</td>\n",
       "      <td>-0.086940</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AACS</td>\n",
       "      <td>ENSG00000081760</td>\n",
       "      <td>[Enzymes, Predictedmembraneproteins]</td>\n",
       "      <td>[Fattyacidmetabolism, Lipidmetabolism]</td>\n",
       "      <td>[Ligase]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[Vesicles]</td>\n",
       "      <td>-0.469481</td>\n",
       "      <td>-0.073416</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AADAC</td>\n",
       "      <td>ENSG00000114771</td>\n",
       "      <td>[Enzymes, Predictedintracellularproteins]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[Hydrolase]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>-0.634534</td>\n",
       "      <td>-0.105024</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15008</th>\n",
       "      <td>ZXDB</td>\n",
       "      <td>ENSG00000198455</td>\n",
       "      <td>[Predictedintracellularproteins, Transcription...</td>\n",
       "      <td>[Transcription, Transcriptionregulation]</td>\n",
       "      <td>[Activator]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[Nucleoplasm]</td>\n",
       "      <td>-0.603586</td>\n",
       "      <td>-0.095432</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15009</th>\n",
       "      <td>ZXDC</td>\n",
       "      <td>ENSG00000070476</td>\n",
       "      <td>[Predictedintracellularproteins, Transcription...</td>\n",
       "      <td>[Transcription, Transcriptionregulation]</td>\n",
       "      <td>[Activator]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[Nucleoli]</td>\n",
       "      <td>-0.175480</td>\n",
       "      <td>-0.095746</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15010</th>\n",
       "      <td>ZYG11B</td>\n",
       "      <td>ENSG00000162378</td>\n",
       "      <td>[Predictedintracellularproteins]</td>\n",
       "      <td>[Ublconjugationpathway]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[Golgiapparatus, Intermediatefilaments]</td>\n",
       "      <td>-0.325059</td>\n",
       "      <td>-0.087883</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15011</th>\n",
       "      <td>ZYX</td>\n",
       "      <td>ENSG00000159840</td>\n",
       "      <td>[Plasmaproteins, Predictedintracellularproteins]</td>\n",
       "      <td>[Celladhesion, Host-virusinteraction]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[Plasmamembrane, Actinfilaments, Focaladhesion...</td>\n",
       "      <td>2.001158</td>\n",
       "      <td>-0.073730</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15012</th>\n",
       "      <td>ZZEF1</td>\n",
       "      <td>ENSG00000074755</td>\n",
       "      <td>[Predictedmembraneproteins]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[Nucleoplasm, Mitochondria]</td>\n",
       "      <td>-0.098111</td>\n",
       "      <td>-0.057218</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15013 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Gene          Ensembl  \\\n",
       "0         A2M  ENSG00000175899   \n",
       "1      A4GALT  ENSG00000128274   \n",
       "2        AAAS  ENSG00000094914   \n",
       "3        AACS  ENSG00000081760   \n",
       "4       AADAC  ENSG00000114771   \n",
       "...       ...              ...   \n",
       "15008    ZXDB  ENSG00000198455   \n",
       "15009    ZXDC  ENSG00000070476   \n",
       "15010  ZYG11B  ENSG00000162378   \n",
       "15011     ZYX  ENSG00000159840   \n",
       "15012   ZZEF1  ENSG00000074755   \n",
       "\n",
       "                                           Protein class  \\\n",
       "0      [Cancer-relatedgenes, Candidatecardiovasculard...   \n",
       "1                   [Enzymes, Predictedmembraneproteins]   \n",
       "2      [Diseaserelatedgenes, Potentialdrugtargets, Pr...   \n",
       "3                   [Enzymes, Predictedmembraneproteins]   \n",
       "4              [Enzymes, Predictedintracellularproteins]   \n",
       "...                                                  ...   \n",
       "15008  [Predictedintracellularproteins, Transcription...   \n",
       "15009  [Predictedintracellularproteins, Transcription...   \n",
       "15010                   [Predictedintracellularproteins]   \n",
       "15011   [Plasmaproteins, Predictedintracellularproteins]   \n",
       "15012                        [Predictedmembraneproteins]   \n",
       "\n",
       "                                      Biological process  \\\n",
       "0                                                     []   \n",
       "1                   [Lipidbiosynthesis, Lipidmetabolism]   \n",
       "2      [mRNAtransport, Proteintransport, Translocatio...   \n",
       "3                 [Fattyacidmetabolism, Lipidmetabolism]   \n",
       "4                                                     []   \n",
       "...                                                  ...   \n",
       "15008           [Transcription, Transcriptionregulation]   \n",
       "15009           [Transcription, Transcriptionregulation]   \n",
       "15010                            [Ublconjugationpathway]   \n",
       "15011              [Celladhesion, Host-virusinteraction]   \n",
       "15012                                                 []   \n",
       "\n",
       "                                 Molecular function    Disease involvement  \\\n",
       "0      [Proteaseinhibitor, Serineproteaseinhibitor]  [Cancer-relatedgenes]   \n",
       "1                [Glycosyltransferase, Transferase]                     []   \n",
       "2                                                []      [Diseasemutation]   \n",
       "3                                          [Ligase]                     []   \n",
       "4                                       [Hydrolase]                     []   \n",
       "...                                             ...                    ...   \n",
       "15008                                   [Activator]                     []   \n",
       "15009                                   [Activator]                     []   \n",
       "15010                                            []                     []   \n",
       "15011                                            []                     []   \n",
       "15012                                            []                     []   \n",
       "\n",
       "                                    Subcellular location  \\\n",
       "0                                                     []   \n",
       "1                                         [Mitochondria]   \n",
       "2                 [Nuclearmembrane, Centrosome, Cytosol]   \n",
       "3                                             [Vesicles]   \n",
       "4                                                     []   \n",
       "...                                                  ...   \n",
       "15008                                      [Nucleoplasm]   \n",
       "15009                                         [Nucleoli]   \n",
       "15010            [Golgiapparatus, Intermediatefilaments]   \n",
       "15011  [Plasmamembrane, Actinfilaments, Focaladhesion...   \n",
       "15012                        [Nucleoplasm, Mitochondria]   \n",
       "\n",
       "       Tissue RNA - lung [NX]  \\\n",
       "0                   11.006867   \n",
       "1                   -0.000111   \n",
       "2                   -0.031058   \n",
       "3                   -0.469481   \n",
       "4                   -0.634534   \n",
       "...                       ...   \n",
       "15008               -0.603586   \n",
       "15009               -0.175480   \n",
       "15010               -0.325059   \n",
       "15011                2.001158   \n",
       "15012               -0.098111   \n",
       "\n",
       "       Single Cell Type RNA - Mucus-secreting cells [NX]  Score_gda  \n",
       "0                                              -0.105024          0  \n",
       "1                                              -0.105024          0  \n",
       "2                                              -0.086940          0  \n",
       "3                                              -0.073416          0  \n",
       "4                                              -0.105024          0  \n",
       "...                                                  ...        ...  \n",
       "15008                                          -0.095432          0  \n",
       "15009                                          -0.095746          0  \n",
       "15010                                          -0.087883          0  \n",
       "15011                                          -0.073730          0  \n",
       "15012                                          -0.057218          0  \n",
       "\n",
       "[15013 rows x 10 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Clean data\n",
    "def explode(feature) :\n",
    "    return feature.apply(lambda x: x.replace(' ', '').split(','))\n",
    "\n",
    "hpa_clean = hpa_label_data.fillna('')\n",
    "for ft in discrete_features :\n",
    "    hpa_clean[ft] = explode(hpa_clean[ft])\n",
    "hpa_clean"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Collect all discrete Features\n",
    "protein_class = hpa_clean[\"Protein class\"].explode().unique()\n",
    "biological_process = hpa_clean[\"Biological process\"].explode().unique()\n",
    "molecular_function = hpa_clean[\"Molecular function\"].explode().unique()\n",
    "disease_involvement = hpa_clean[\"Disease involvement\"].explode().unique()\n",
    "subcellular_location = hpa_clean[\"Subcellular location\"].explode().unique()\n",
    "GO_features = np.concatenate([protein_class, biological_process, molecular_function, disease_involvement, subcellular_location])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Cancer-relatedgenes</th>\n",
       "      <th>Candidatecardiovasculardiseasegenes</th>\n",
       "      <th>Plasmaproteins</th>\n",
       "      <th>Predictedsecretedproteins</th>\n",
       "      <th>Enzymes</th>\n",
       "      <th>Predictedmembraneproteins</th>\n",
       "      <th>Diseaserelatedgenes</th>\n",
       "      <th>Potentialdrugtargets</th>\n",
       "      <th>Predictedintracellularproteins</th>\n",
       "      <th>Transporters</th>\n",
       "      <th>...</th>\n",
       "      <th>Mitoticchromosome</th>\n",
       "      <th>Intermediatefilaments</th>\n",
       "      <th>Cytokineticbridge</th>\n",
       "      <th>Rods&amp;Rings</th>\n",
       "      <th>Mitoticspindle</th>\n",
       "      <th>Aggresome</th>\n",
       "      <th>Lysosomes</th>\n",
       "      <th>Microtubuleends</th>\n",
       "      <th>Kinetochore</th>\n",
       "      <th>Cleavagefurrow</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A4GALT</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAS</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AACS</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AADAC</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZXDB</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZXDC</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11B</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15013 rows × 556 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Cancer-relatedgenes  Candidatecardiovasculardiseasegenes  \\\n",
       "Gene                                                               \n",
       "A2M                       1                                    1   \n",
       "A4GALT                    0                                    0   \n",
       "AAAS                      0                                    0   \n",
       "AACS                      0                                    0   \n",
       "AADAC                     0                                    0   \n",
       "...                     ...                                  ...   \n",
       "ZXDB                      0                                    0   \n",
       "ZXDC                      0                                    0   \n",
       "ZYG11B                    0                                    0   \n",
       "ZYX                       0                                    0   \n",
       "ZZEF1                     0                                    0   \n",
       "\n",
       "        Plasmaproteins  Predictedsecretedproteins  Enzymes  \\\n",
       "Gene                                                         \n",
       "A2M                  1                          1        0   \n",
       "A4GALT               0                          0        1   \n",
       "AAAS                 0                          0        0   \n",
       "AACS                 0                          0        1   \n",
       "AADAC                0                          0        1   \n",
       "...                ...                        ...      ...   \n",
       "ZXDB                 0                          0        0   \n",
       "ZXDC                 0                          0        0   \n",
       "ZYG11B               0                          0        0   \n",
       "ZYX                  1                          0        0   \n",
       "ZZEF1                0                          0        0   \n",
       "\n",
       "        Predictedmembraneproteins  Diseaserelatedgenes  Potentialdrugtargets  \\\n",
       "Gene                                                                           \n",
       "A2M                             0                    0                     0   \n",
       "A4GALT                          1                    0                     0   \n",
       "AAAS                            0                    1                     1   \n",
       "AACS                            1                    0                     0   \n",
       "AADAC                           0                    0                     0   \n",
       "...                           ...                  ...                   ...   \n",
       "ZXDB                            0                    0                     0   \n",
       "ZXDC                            0                    0                     0   \n",
       "ZYG11B                          0                    0                     0   \n",
       "ZYX                             0                    0                     0   \n",
       "ZZEF1                           1                    0                     0   \n",
       "\n",
       "        Predictedintracellularproteins  Transporters  ...  Mitoticchromosome  \\\n",
       "Gene                                                  ...                      \n",
       "A2M                                  0             0  ...                  0   \n",
       "A4GALT                               0             0  ...                  0   \n",
       "AAAS                                 1             1  ...                  0   \n",
       "AACS                                 0             0  ...                  0   \n",
       "AADAC                                1             0  ...                  0   \n",
       "...                                ...           ...  ...                ...   \n",
       "ZXDB                                 1             0  ...                  0   \n",
       "ZXDC                                 1             0  ...                  0   \n",
       "ZYG11B                               1             0  ...                  0   \n",
       "ZYX                                  1             0  ...                  0   \n",
       "ZZEF1                                0             0  ...                  0   \n",
       "\n",
       "        Intermediatefilaments  Cytokineticbridge  Rods&Rings  Mitoticspindle  \\\n",
       "Gene                                                                           \n",
       "A2M                         0                  0           0               0   \n",
       "A4GALT                      0                  0           0               0   \n",
       "AAAS                        0                  0           0               0   \n",
       "AACS                        0                  0           0               0   \n",
       "AADAC                       0                  0           0               0   \n",
       "...                       ...                ...         ...             ...   \n",
       "ZXDB                        0                  0           0               0   \n",
       "ZXDC                        0                  0           0               0   \n",
       "ZYG11B                      1                  0           0               0   \n",
       "ZYX                         0                  0           0               0   \n",
       "ZZEF1                       0                  0           0               0   \n",
       "\n",
       "        Aggresome  Lysosomes  Microtubuleends  Kinetochore  Cleavagefurrow  \n",
       "Gene                                                                        \n",
       "A2M             0          0                0            0               0  \n",
       "A4GALT          0          0                0            0               0  \n",
       "AAAS            0          0                0            0               0  \n",
       "AACS            0          0                0            0               0  \n",
       "AADAC           0          0                0            0               0  \n",
       "...           ...        ...              ...          ...             ...  \n",
       "ZXDB            0          0                0            0               0  \n",
       "ZXDC            0          0                0            0               0  \n",
       "ZYG11B          0          0                0            0               0  \n",
       "ZYX             0          0                0            0               0  \n",
       "ZZEF1           0          0                0            0               0  \n",
       "\n",
       "[15013 rows x 556 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# One hot encode discrete features\n",
    "RowFeatures = pd.DataFrame(data = 0,index = hpa_clean['Gene'],columns=GO_features)\n",
    "counter = 0\n",
    "for index, row in RowFeatures.iterrows() :\n",
    "    features = hpa_clean.loc[counter][['Protein class', 'Biological process', 'Molecular function', 'Disease involvement', 'Subcellular location']].to_list()\n",
    "    flattened = [item for sublist in features for item in sublist if item]\n",
    "    for t in flattened :\n",
    "        row[t] = 1\n",
    "    counter +=1\n",
    "RowFeatures"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Matrix Factorization + Optimization\n",
    "Using SVD to factorize our discrete variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30\n",
      "40\n",
      "50\n",
      "60\n",
      "70\n",
      "80\n",
      "90\n",
      "100\n",
      "110\n"
     ]
    }
   ],
   "source": [
    "from sklearn.decomposition import TruncatedSVD\n",
    "\n",
    "#dictionary of data frames\n",
    "feat_embeds = {}\n",
    "\n",
    "#Create the different feature embeddings\n",
    "# try: 30,40,50,60,70,80,90,100,110\n",
    "for i in range(9) :\n",
    "    n_comp = (i+3)*10\n",
    "    svd = TruncatedSVD(n_components = n_comp)\n",
    "    svdModel = svd.fit(RowFeatures)\n",
    "    visits_emb = svdModel.transform(RowFeatures)\n",
    "    visits_emb_df = pd.DataFrame(data=visits_emb, index=RowFeatures.index)\n",
    "    feat_embeds[n_comp] = pd.merge(visits_emb_df, hpa_clean[identifiers+continuous_features+['Score_gda']], on='Gene',how='inner')\n",
    "    \n",
    "    print(n_comp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.metrics import classification_report\n",
    "\n",
    "def eval_model(model, X, y):\n",
    "    preds = model.predict(X)\n",
    "    clf_report = classification_report(y, preds, labels=[0, 1], target_names=['negative', 'positive'], digits=2)\n",
    "    print(clf_report)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Int64Index([ 6303, 13434,  7844, 13458, 11123,  5170,  2553,  4224,  4787,\n",
       "             5202,\n",
       "            ...\n",
       "            11464, 14174,  7485, 13635,  4161,  6844,  3151,  6984,  3160,\n",
       "             2851],\n",
       "           dtype='int64', length=400)"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Negative sample\n",
    "negIDs = hpa_label_data.loc[hpa_label_data.Score_gda == 0].sample(hpa_label_data.Score_gda.sum()).index\n",
    "negIDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "----------- Number of embeddings:  30\n",
      "Best Parameters\n",
      "{'criterion': 'gini', 'max_depth': 3, 'n_estimators': 50}\n",
      "\n",
      "\n",
      "Training Metrics\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "    negative       0.76      0.86      0.81       300\n",
      "    positive       0.84      0.73      0.78       300\n",
      "\n",
      "    accuracy                           0.80       600\n",
      "   macro avg       0.80      0.80      0.80       600\n",
      "weighted avg       0.80      0.80      0.80       600\n",
      "\n",
      "\n",
      "Testing Metrics\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "    negative       0.66      0.76      0.71       100\n",
      "    positive       0.72      0.61      0.66       100\n",
      "\n",
      "    accuracy                           0.69       200\n",
      "   macro avg       0.69      0.69      0.68       200\n",
      "weighted avg       0.69      0.69      0.68       200\n",
      "\n",
      "------------------------------------------\n",
      "----------- Number of embeddings:  40\n",
      "Best Parameters\n",
      "{'criterion': 'gini', 'max_depth': 2, 'n_estimators': 50}\n",
      "\n",
      "\n",
      "Training Metrics\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "    negative       0.72      0.86      0.78       300\n",
      "    positive       0.82      0.67      0.74       300\n",
      "\n",
      "    accuracy                           0.76       600\n",
      "   macro avg       0.77      0.76      0.76       600\n",
      "weighted avg       0.77      0.76      0.76       600\n",
      "\n",
      "\n",
      "Testing Metrics\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "    negative       0.69      0.85      0.76       100\n",
      "    positive       0.80      0.61      0.69       100\n",
      "\n",
      "    accuracy                           0.73       200\n",
      "   macro avg       0.74      0.73      0.73       200\n",
      "weighted avg       0.74      0.73      0.73       200\n",
      "\n",
      "------------------------------------------\n",
      "----------- Number of embeddings:  50\n",
      "Best Parameters\n",
      "{'criterion': 'gini', 'max_depth': 4, 'n_estimators': 150}\n",
      "\n",
      "\n",
      "Training Metrics\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "    negative       0.79      0.89      0.84       300\n",
      "    positive       0.88      0.76      0.82       300\n",
      "\n",
      "    accuracy                           0.83       600\n",
      "   macro avg       0.83      0.83      0.83       600\n",
      "weighted avg       0.83      0.83      0.83       600\n",
      "\n",
      "\n",
      "Testing Metrics\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "    negative       0.67      0.80      0.73       100\n",
      "    positive       0.75      0.61      0.67       100\n",
      "\n",
      "    accuracy                           0.70       200\n",
      "   macro avg       0.71      0.71      0.70       200\n",
      "weighted avg       0.71      0.70      0.70       200\n",
      "\n",
      "------------------------------------------\n",
      "----------- Number of embeddings:  60\n",
      "Best Parameters\n",
      "{'criterion': 'gini', 'max_depth': 5, 'n_estimators': 200}\n",
      "\n",
      "\n",
      "Training Metrics\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "    negative       0.83      0.95      0.89       300\n",
      "    positive       0.95      0.81      0.87       300\n",
      "\n",
      "    accuracy                           0.88       600\n",
      "   macro avg       0.89      0.88      0.88       600\n",
      "weighted avg       0.89      0.88      0.88       600\n",
      "\n",
      "\n",
      "Testing Metrics\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "    negative       0.68      0.81      0.74       100\n",
      "    positive       0.76      0.61      0.68       100\n",
      "\n",
      "    accuracy                           0.71       200\n",
      "   macro avg       0.72      0.71      0.71       200\n",
      "weighted avg       0.72      0.71      0.71       200\n",
      "\n",
      "------------------------------------------\n",
      "----------- Number of embeddings:  70\n",
      "Best Parameters\n",
      "{'criterion': 'gini', 'max_depth': 4, 'n_estimators': 150}\n",
      "\n",
      "\n",
      "Training Metrics\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "    negative       0.80      0.91      0.85       300\n",
      "    positive       0.89      0.77      0.83       300\n",
      "\n",
      "    accuracy                           0.84       600\n",
      "   macro avg       0.84      0.84      0.84       600\n",
      "weighted avg       0.84      0.84      0.84       600\n",
      "\n",
      "\n",
      "Testing Metrics\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "    negative       0.69      0.81      0.75       100\n",
      "    positive       0.77      0.64      0.70       100\n",
      "\n",
      "    accuracy                           0.73       200\n",
      "   macro avg       0.73      0.73      0.72       200\n",
      "weighted avg       0.73      0.72      0.72       200\n",
      "\n",
      "------------------------------------------\n",
      "----------- Number of embeddings:  80\n",
      "Best Parameters\n",
      "{'criterion': 'gini', 'max_depth': 2, 'n_estimators': 150}\n",
      "\n",
      "\n",
      "Training Metrics\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "    negative       0.74      0.86      0.80       300\n",
      "    positive       0.83      0.71      0.76       300\n",
      "\n",
      "    accuracy                           0.78       600\n",
      "   macro avg       0.79      0.78      0.78       600\n",
      "weighted avg       0.79      0.78      0.78       600\n",
      "\n",
      "\n",
      "Testing Metrics\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "    negative       0.68      0.80      0.73       100\n",
      "    positive       0.76      0.62      0.68       100\n",
      "\n",
      "    accuracy                           0.71       200\n",
      "   macro avg       0.72      0.71      0.71       200\n",
      "weighted avg       0.72      0.71      0.71       200\n",
      "\n",
      "------------------------------------------\n",
      "----------- Number of embeddings:  90\n",
      "Best Parameters\n",
      "{'criterion': 'gini', 'max_depth': 5, 'n_estimators': 150}\n",
      "\n",
      "\n",
      "Training Metrics\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "    negative       0.85      0.98      0.91       300\n",
      "    positive       0.98      0.82      0.89       300\n",
      "\n",
      "    accuracy                           0.90       600\n",
      "   macro avg       0.91      0.90      0.90       600\n",
      "weighted avg       0.91      0.90      0.90       600\n",
      "\n",
      "\n",
      "Testing Metrics\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "    negative       0.68      0.78      0.73       100\n",
      "    positive       0.74      0.63      0.68       100\n",
      "\n",
      "    accuracy                           0.70       200\n",
      "   macro avg       0.71      0.71      0.70       200\n",
      "weighted avg       0.71      0.70      0.70       200\n",
      "\n",
      "------------------------------------------\n",
      "----------- Number of embeddings:  100\n",
      "Best Parameters\n",
      "{'criterion': 'gini', 'max_depth': 5, 'n_estimators': 150}\n",
      "\n",
      "\n",
      "Training Metrics\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "    negative       0.87      0.96      0.91       300\n",
      "    positive       0.96      0.85      0.90       300\n",
      "\n",
      "    accuracy                           0.91       600\n",
      "   macro avg       0.91      0.91      0.91       600\n",
      "weighted avg       0.91      0.91      0.91       600\n",
      "\n",
      "\n",
      "Testing Metrics\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "    negative       0.68      0.82      0.75       100\n",
      "    positive       0.78      0.62      0.69       100\n",
      "\n",
      "    accuracy                           0.72       200\n",
      "   macro avg       0.73      0.72      0.72       200\n",
      "weighted avg       0.73      0.72      0.72       200\n",
      "\n",
      "------------------------------------------\n",
      "----------- Number of embeddings:  110\n",
      "Best Parameters\n",
      "{'criterion': 'gini', 'max_depth': 3, 'n_estimators': 150}\n",
      "\n",
      "\n",
      "Training Metrics\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "    negative       0.79      0.90      0.84       300\n",
      "    positive       0.88      0.76      0.82       300\n",
      "\n",
      "    accuracy                           0.83       600\n",
      "   macro avg       0.84      0.83      0.83       600\n",
      "weighted avg       0.84      0.83      0.83       600\n",
      "\n",
      "\n",
      "Testing Metrics\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "    negative       0.69      0.76      0.72       100\n",
      "    positive       0.73      0.66      0.69       100\n",
      "\n",
      "    accuracy                           0.71       200\n",
      "   macro avg       0.71      0.71      0.71       200\n",
      "weighted avg       0.71      0.71      0.71       200\n",
      "\n",
      "------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "for i in range(7,8):\n",
    "    n_comp = (i+3)*10\n",
    "    \n",
    "    # Create Sample data\n",
    "    data = feat_embeds[n_comp]\n",
    "    data_sample = data.loc[ data.index.isin(negIDs) | data.Score_gda == 1]\n",
    "    \n",
    "    # separate features and labels\n",
    "    node_feats = data_sample.iloc[:,1:-1]\n",
    "    node_labels = data_sample.iloc[:,-1]\n",
    "    \n",
    "    # create train-test split\n",
    "    from sklearn.model_selection import train_test_split\n",
    "    test_size = 0.25\n",
    "\n",
    "    X_train, X_test, y_train, y_test = train_test_split(node_feats, node_labels, test_size=test_size, shuffle=True, stratify=node_labels, random_state=360)\n",
    "    # NOTE: train test split is shuffled and stratified across labels\n",
    "\n",
    "\n",
    "    from sklearn.ensemble import RandomForestClassifier\n",
    "    from sklearn.model_selection import GridSearchCV\n",
    "\n",
    "    # parameter grid to search over\n",
    "    parameters = {\n",
    "                    'n_estimators':[20, 50, 100, 150, 200],\n",
    "                    'criterion':['gini'],\n",
    "                    'max_depth':[1, 2, 3, 4, 5]\n",
    "                 }\n",
    "\n",
    "    # base random forest model\n",
    "    rf = RandomForestClassifier(n_jobs=-1)\n",
    "\n",
    "    # perform a gridsearch with 5-fold crossvalidation to find the best model\n",
    "    rf_clf = GridSearchCV(rf, parameters, n_jobs=-1, refit=True, cv=5, return_train_score=True)\n",
    "\n",
    "    rf_clf.fit(X_train, y_train)\n",
    "\n",
    "    print('----------- Number of embeddings: ',n_comp)\n",
    "    # show choice of parameters that yielded the best performance\n",
    "    print('Best Parameters')\n",
    "    print(rf_clf.best_params_)\n",
    "    print('\\n')\n",
    "\n",
    "    # evaluate model\n",
    "    print('Training Metrics')\n",
    "    eval_model(rf_clf.best_estimator_, X_train, y_train)\n",
    "\n",
    "    print()\n",
    "    print('Testing Metrics')\n",
    "    eval_model(rf_clf.best_estimator_, X_test, y_test)\n",
    "    print('------------------------------------------')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Ensembl</th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>...</th>\n",
       "      <th>92</th>\n",
       "      <th>93</th>\n",
       "      <th>94</th>\n",
       "      <th>95</th>\n",
       "      <th>96</th>\n",
       "      <th>97</th>\n",
       "      <th>98</th>\n",
       "      <th>99</th>\n",
       "      <th>Tissue RNA - lung [NX]</th>\n",
       "      <th>Single Cell Type RNA - Mucus-secreting cells [NX]</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000175899</td>\n",
       "      <td>0.443505</td>\n",
       "      <td>0.313574</td>\n",
       "      <td>0.430307</td>\n",
       "      <td>-0.285337</td>\n",
       "      <td>-0.351412</td>\n",
       "      <td>1.545059</td>\n",
       "      <td>-0.450936</td>\n",
       "      <td>-0.444200</td>\n",
       "      <td>-0.116774</td>\n",
       "      <td>...</td>\n",
       "      <td>0.033076</td>\n",
       "      <td>0.069726</td>\n",
       "      <td>0.159449</td>\n",
       "      <td>0.031158</td>\n",
       "      <td>0.087127</td>\n",
       "      <td>0.812502</td>\n",
       "      <td>0.652882</td>\n",
       "      <td>0.117809</td>\n",
       "      <td>11.006867</td>\n",
       "      <td>-0.105024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000128274</td>\n",
       "      <td>0.532289</td>\n",
       "      <td>0.789331</td>\n",
       "      <td>-0.122159</td>\n",
       "      <td>0.019000</td>\n",
       "      <td>1.427162</td>\n",
       "      <td>-0.065659</td>\n",
       "      <td>0.001669</td>\n",
       "      <td>0.386983</td>\n",
       "      <td>-0.028900</td>\n",
       "      <td>...</td>\n",
       "      <td>0.283313</td>\n",
       "      <td>0.104073</td>\n",
       "      <td>0.122308</td>\n",
       "      <td>-0.010488</td>\n",
       "      <td>0.071484</td>\n",
       "      <td>-0.042168</td>\n",
       "      <td>0.138073</td>\n",
       "      <td>-0.061587</td>\n",
       "      <td>-0.000111</td>\n",
       "      <td>-0.105024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000094914</td>\n",
       "      <td>1.605229</td>\n",
       "      <td>0.936681</td>\n",
       "      <td>0.345886</td>\n",
       "      <td>0.183699</td>\n",
       "      <td>-1.001764</td>\n",
       "      <td>-0.846346</td>\n",
       "      <td>0.550995</td>\n",
       "      <td>0.368205</td>\n",
       "      <td>0.202731</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.193989</td>\n",
       "      <td>0.169840</td>\n",
       "      <td>-0.373577</td>\n",
       "      <td>0.174107</td>\n",
       "      <td>-0.118314</td>\n",
       "      <td>0.093455</td>\n",
       "      <td>0.209565</td>\n",
       "      <td>-0.057426</td>\n",
       "      <td>-0.031058</td>\n",
       "      <td>-0.086940</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000081760</td>\n",
       "      <td>0.467196</td>\n",
       "      <td>0.690415</td>\n",
       "      <td>-0.066723</td>\n",
       "      <td>0.335478</td>\n",
       "      <td>1.017124</td>\n",
       "      <td>0.029579</td>\n",
       "      <td>-0.131714</td>\n",
       "      <td>0.212387</td>\n",
       "      <td>0.075120</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.350115</td>\n",
       "      <td>-0.241371</td>\n",
       "      <td>0.016416</td>\n",
       "      <td>-0.012377</td>\n",
       "      <td>0.039609</td>\n",
       "      <td>-0.034740</td>\n",
       "      <td>0.007967</td>\n",
       "      <td>0.039796</td>\n",
       "      <td>-0.469481</td>\n",
       "      <td>-0.073416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000114771</td>\n",
       "      <td>0.956718</td>\n",
       "      <td>0.202132</td>\n",
       "      <td>-0.490047</td>\n",
       "      <td>-0.524460</td>\n",
       "      <td>0.467132</td>\n",
       "      <td>-0.207775</td>\n",
       "      <td>-0.494652</td>\n",
       "      <td>0.192615</td>\n",
       "      <td>0.077836</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001733</td>\n",
       "      <td>-0.000615</td>\n",
       "      <td>-0.004378</td>\n",
       "      <td>-0.000040</td>\n",
       "      <td>0.002857</td>\n",
       "      <td>-0.010442</td>\n",
       "      <td>0.016493</td>\n",
       "      <td>0.005563</td>\n",
       "      <td>-0.634534</td>\n",
       "      <td>-0.105024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15008</th>\n",
       "      <td>ENSG00000198455</td>\n",
       "      <td>1.436813</td>\n",
       "      <td>-1.455513</td>\n",
       "      <td>0.510199</td>\n",
       "      <td>0.224208</td>\n",
       "      <td>0.323736</td>\n",
       "      <td>-0.171141</td>\n",
       "      <td>-0.060697</td>\n",
       "      <td>0.088626</td>\n",
       "      <td>0.103071</td>\n",
       "      <td>...</td>\n",
       "      <td>0.003771</td>\n",
       "      <td>0.063757</td>\n",
       "      <td>-0.002031</td>\n",
       "      <td>0.008038</td>\n",
       "      <td>-0.023358</td>\n",
       "      <td>0.013035</td>\n",
       "      <td>-0.001910</td>\n",
       "      <td>0.022492</td>\n",
       "      <td>-0.603586</td>\n",
       "      <td>-0.095432</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15009</th>\n",
       "      <td>ENSG00000070476</td>\n",
       "      <td>1.140170</td>\n",
       "      <td>-1.162868</td>\n",
       "      <td>0.551409</td>\n",
       "      <td>0.127004</td>\n",
       "      <td>0.119115</td>\n",
       "      <td>-0.120434</td>\n",
       "      <td>-0.421260</td>\n",
       "      <td>0.798186</td>\n",
       "      <td>0.041917</td>\n",
       "      <td>...</td>\n",
       "      <td>0.005172</td>\n",
       "      <td>0.045257</td>\n",
       "      <td>0.005712</td>\n",
       "      <td>0.001190</td>\n",
       "      <td>-0.027927</td>\n",
       "      <td>0.016542</td>\n",
       "      <td>-0.008024</td>\n",
       "      <td>0.035935</td>\n",
       "      <td>-0.175480</td>\n",
       "      <td>-0.095746</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15010</th>\n",
       "      <td>ENSG00000162378</td>\n",
       "      <td>0.782655</td>\n",
       "      <td>-0.057857</td>\n",
       "      <td>-0.397211</td>\n",
       "      <td>0.017498</td>\n",
       "      <td>-0.135123</td>\n",
       "      <td>-0.248882</td>\n",
       "      <td>-0.438176</td>\n",
       "      <td>0.113363</td>\n",
       "      <td>-0.295955</td>\n",
       "      <td>...</td>\n",
       "      <td>0.028633</td>\n",
       "      <td>-0.015439</td>\n",
       "      <td>-0.001466</td>\n",
       "      <td>-0.019047</td>\n",
       "      <td>-0.041910</td>\n",
       "      <td>0.019608</td>\n",
       "      <td>0.020722</td>\n",
       "      <td>-0.070376</td>\n",
       "      <td>-0.325059</td>\n",
       "      <td>-0.087883</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15011</th>\n",
       "      <td>ENSG00000159840</td>\n",
       "      <td>1.042306</td>\n",
       "      <td>0.183392</td>\n",
       "      <td>-0.327697</td>\n",
       "      <td>0.065425</td>\n",
       "      <td>-0.459893</td>\n",
       "      <td>0.507171</td>\n",
       "      <td>-0.610148</td>\n",
       "      <td>0.067066</td>\n",
       "      <td>0.368216</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.135105</td>\n",
       "      <td>0.034374</td>\n",
       "      <td>0.453396</td>\n",
       "      <td>0.068095</td>\n",
       "      <td>-0.162314</td>\n",
       "      <td>-0.099047</td>\n",
       "      <td>0.128805</td>\n",
       "      <td>-0.140204</td>\n",
       "      <td>2.001158</td>\n",
       "      <td>-0.073730</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15012</th>\n",
       "      <td>ENSG00000074755</td>\n",
       "      <td>0.553160</td>\n",
       "      <td>0.068709</td>\n",
       "      <td>0.013208</td>\n",
       "      <td>0.760541</td>\n",
       "      <td>0.592771</td>\n",
       "      <td>-0.094544</td>\n",
       "      <td>0.326136</td>\n",
       "      <td>-0.636445</td>\n",
       "      <td>0.047971</td>\n",
       "      <td>...</td>\n",
       "      <td>0.005748</td>\n",
       "      <td>0.011550</td>\n",
       "      <td>0.000926</td>\n",
       "      <td>0.015975</td>\n",
       "      <td>0.003848</td>\n",
       "      <td>0.001315</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>-0.013763</td>\n",
       "      <td>-0.098111</td>\n",
       "      <td>-0.057218</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15013 rows × 103 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               Ensembl         0         1         2         3         4  \\\n",
       "0      ENSG00000175899  0.443505  0.313574  0.430307 -0.285337 -0.351412   \n",
       "1      ENSG00000128274  0.532289  0.789331 -0.122159  0.019000  1.427162   \n",
       "2      ENSG00000094914  1.605229  0.936681  0.345886  0.183699 -1.001764   \n",
       "3      ENSG00000081760  0.467196  0.690415 -0.066723  0.335478  1.017124   \n",
       "4      ENSG00000114771  0.956718  0.202132 -0.490047 -0.524460  0.467132   \n",
       "...                ...       ...       ...       ...       ...       ...   \n",
       "15008  ENSG00000198455  1.436813 -1.455513  0.510199  0.224208  0.323736   \n",
       "15009  ENSG00000070476  1.140170 -1.162868  0.551409  0.127004  0.119115   \n",
       "15010  ENSG00000162378  0.782655 -0.057857 -0.397211  0.017498 -0.135123   \n",
       "15011  ENSG00000159840  1.042306  0.183392 -0.327697  0.065425 -0.459893   \n",
       "15012  ENSG00000074755  0.553160  0.068709  0.013208  0.760541  0.592771   \n",
       "\n",
       "              5         6         7         8  ...        92        93  \\\n",
       "0      1.545059 -0.450936 -0.444200 -0.116774  ...  0.033076  0.069726   \n",
       "1     -0.065659  0.001669  0.386983 -0.028900  ...  0.283313  0.104073   \n",
       "2     -0.846346  0.550995  0.368205  0.202731  ... -0.193989  0.169840   \n",
       "3      0.029579 -0.131714  0.212387  0.075120  ... -0.350115 -0.241371   \n",
       "4     -0.207775 -0.494652  0.192615  0.077836  ...  0.001733 -0.000615   \n",
       "...         ...       ...       ...       ...  ...       ...       ...   \n",
       "15008 -0.171141 -0.060697  0.088626  0.103071  ...  0.003771  0.063757   \n",
       "15009 -0.120434 -0.421260  0.798186  0.041917  ...  0.005172  0.045257   \n",
       "15010 -0.248882 -0.438176  0.113363 -0.295955  ...  0.028633 -0.015439   \n",
       "15011  0.507171 -0.610148  0.067066  0.368216  ... -0.135105  0.034374   \n",
       "15012 -0.094544  0.326136 -0.636445  0.047971  ...  0.005748  0.011550   \n",
       "\n",
       "             94        95        96        97        98        99  \\\n",
       "0      0.159449  0.031158  0.087127  0.812502  0.652882  0.117809   \n",
       "1      0.122308 -0.010488  0.071484 -0.042168  0.138073 -0.061587   \n",
       "2     -0.373577  0.174107 -0.118314  0.093455  0.209565 -0.057426   \n",
       "3      0.016416 -0.012377  0.039609 -0.034740  0.007967  0.039796   \n",
       "4     -0.004378 -0.000040  0.002857 -0.010442  0.016493  0.005563   \n",
       "...         ...       ...       ...       ...       ...       ...   \n",
       "15008 -0.002031  0.008038 -0.023358  0.013035 -0.001910  0.022492   \n",
       "15009  0.005712  0.001190 -0.027927  0.016542 -0.008024  0.035935   \n",
       "15010 -0.001466 -0.019047 -0.041910  0.019608  0.020722 -0.070376   \n",
       "15011  0.453396  0.068095 -0.162314 -0.099047  0.128805 -0.140204   \n",
       "15012  0.000926  0.015975  0.003848  0.001315  0.000012 -0.013763   \n",
       "\n",
       "       Tissue RNA - lung [NX]  \\\n",
       "0                   11.006867   \n",
       "1                   -0.000111   \n",
       "2                   -0.031058   \n",
       "3                   -0.469481   \n",
       "4                   -0.634534   \n",
       "...                       ...   \n",
       "15008               -0.603586   \n",
       "15009               -0.175480   \n",
       "15010               -0.325059   \n",
       "15011                2.001158   \n",
       "15012               -0.098111   \n",
       "\n",
       "       Single Cell Type RNA - Mucus-secreting cells [NX]  \n",
       "0                                              -0.105024  \n",
       "1                                              -0.105024  \n",
       "2                                              -0.086940  \n",
       "3                                              -0.073416  \n",
       "4                                              -0.105024  \n",
       "...                                                  ...  \n",
       "15008                                          -0.095432  \n",
       "15009                                          -0.095746  \n",
       "15010                                          -0.087883  \n",
       "15011                                          -0.073730  \n",
       "15012                                          -0.057218  \n",
       "\n",
       "[15013 rows x 103 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# quick cleanup before saving to csv\n",
    "final_embed = feat_embeds[100]\n",
    "cols = final_embed.columns.tolist()\n",
    "\n",
    "# Replace Score_gda with Label\n",
    "cols.pop(-1)\n",
    "cols.append('Label')\n",
    "final_embed.columns = cols\n",
    "\n",
    "# Rearrange columns\n",
    "cols = [cols[0]] + [cols[101]] + cols[1:101] + cols[102:104]\n",
    "final_embed = final_embed[cols]\n",
    "final_embed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save HPA\n",
    "final_embed.to_csv('../Data/Final/HPA_processed_features_v1.1.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save Gene list\n",
    "final_embed['Ensembl'].to_csv('../Data/Final/HPA_gene_list.tsv', sep='\\t', index=False, header=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
