{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from sklearn.decomposition import TruncatedSVD"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import National Institute of Health Genomic Data Commons (GDA) and Human Protein Atlas (HPA) Data "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl</th>\n",
       "      <th>nih_ssm_in_cohort</th>\n",
       "      <th>nih_ssm_across_gdc</th>\n",
       "      <th>nih_cnv_gain</th>\n",
       "      <th>nih_cnv_loss</th>\n",
       "      <th>nih_tot_mutations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000147481</td>\n",
       "      <td>0.144621</td>\n",
       "      <td>0.028155</td>\n",
       "      <td>0.076023</td>\n",
       "      <td>0.056530</td>\n",
       "      <td>92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000105877</td>\n",
       "      <td>0.144621</td>\n",
       "      <td>0.084772</td>\n",
       "      <td>0.074074</td>\n",
       "      <td>0.007797</td>\n",
       "      <td>105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000188107</td>\n",
       "      <td>0.144621</td>\n",
       "      <td>0.068585</td>\n",
       "      <td>0.111111</td>\n",
       "      <td>0.056530</td>\n",
       "      <td>95</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000125414</td>\n",
       "      <td>0.144621</td>\n",
       "      <td>0.062601</td>\n",
       "      <td>0.005848</td>\n",
       "      <td>0.042885</td>\n",
       "      <td>95</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000009694</td>\n",
       "      <td>0.144621</td>\n",
       "      <td>0.071653</td>\n",
       "      <td>0.052632</td>\n",
       "      <td>0.025341</td>\n",
       "      <td>110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21087</th>\n",
       "      <td>ENSG00000146648</td>\n",
       "      <td>0.146384</td>\n",
       "      <td>0.052244</td>\n",
       "      <td>0.105263</td>\n",
       "      <td>0.019493</td>\n",
       "      <td>51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21088</th>\n",
       "      <td>ENSG00000133958</td>\n",
       "      <td>0.146384</td>\n",
       "      <td>0.059916</td>\n",
       "      <td>0.025341</td>\n",
       "      <td>0.046784</td>\n",
       "      <td>110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21089</th>\n",
       "      <td>ENSG00000134376</td>\n",
       "      <td>0.146384</td>\n",
       "      <td>0.056617</td>\n",
       "      <td>0.085770</td>\n",
       "      <td>0.007797</td>\n",
       "      <td>93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21090</th>\n",
       "      <td>ENSG00000109061</td>\n",
       "      <td>0.146384</td>\n",
       "      <td>0.057691</td>\n",
       "      <td>0.009747</td>\n",
       "      <td>0.040936</td>\n",
       "      <td>93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21091</th>\n",
       "      <td>ENSG00000134516</td>\n",
       "      <td>0.146384</td>\n",
       "      <td>0.054622</td>\n",
       "      <td>0.066277</td>\n",
       "      <td>0.033138</td>\n",
       "      <td>103</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21092 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               ensembl  nih_ssm_in_cohort  nih_ssm_across_gdc  nih_cnv_gain  \\\n",
       "0      ENSG00000147481           0.144621            0.028155      0.076023   \n",
       "1      ENSG00000105877           0.144621            0.084772      0.074074   \n",
       "2      ENSG00000188107           0.144621            0.068585      0.111111   \n",
       "3      ENSG00000125414           0.144621            0.062601      0.005848   \n",
       "4      ENSG00000009694           0.144621            0.071653      0.052632   \n",
       "...                ...                ...                 ...           ...   \n",
       "21087  ENSG00000146648           0.146384            0.052244      0.105263   \n",
       "21088  ENSG00000133958           0.146384            0.059916      0.025341   \n",
       "21089  ENSG00000134376           0.146384            0.056617      0.085770   \n",
       "21090  ENSG00000109061           0.146384            0.057691      0.009747   \n",
       "21091  ENSG00000134516           0.146384            0.054622      0.066277   \n",
       "\n",
       "       nih_cnv_loss  nih_tot_mutations  \n",
       "0          0.056530                 92  \n",
       "1          0.007797                105  \n",
       "2          0.056530                 95  \n",
       "3          0.042885                 95  \n",
       "4          0.025341                110  \n",
       "...             ...                ...  \n",
       "21087      0.019493                 51  \n",
       "21088      0.046784                110  \n",
       "21089      0.007797                 93  \n",
       "21090      0.040936                 93  \n",
       "21091      0.033138                103  \n",
       "\n",
       "[21092 rows x 6 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Process the gdc dataset to include only desired columns and scores\n",
    "gdc = pd.read_csv(\"../data/gdc_luad_genes.csv\")\n",
    "\n",
    "featureNameColumns = ['# SSM Affected Cases in Cohort', '# CNV Gain', '# CNV Loss']\n",
    "for i in featureNameColumns:\n",
    "    gdc[['1', '2', '3', '4']] = gdc[i].replace({',':''}, regex=True).str.split(' ', 3, expand=True)\n",
    "    gdc[i] = gdc['1'].astype(float)/gdc['3'].astype(float)\n",
    "    gdc = gdc.drop(['1', '2', '3', '4'], axis=1)\n",
    "\n",
    "gdc.drop(['Symbol', 'Name', 'Cytoband', 'Type', 'Annotations', 'Survival'], axis=1, inplace=True)\n",
    "\n",
    "gdc[['1', '2', '3']] = gdc['# SSM Affected Cases Across the GDC'].replace({',':''}, regex=True).str.split(' ', 2, expand=True)\n",
    "gdc['# SSM Affected Cases Across the GDC'] = gdc['1'].astype(float)/gdc['3'].astype(float)\n",
    "gdc = gdc.drop(['1', '2', '3'], axis=1)\n",
    "\n",
    "gdc = gdc.rename({'# SSM Affected Cases in Cohort': 'nih_ssm_in_cohort', '# SSM Affected Cases Across the GDC':'nih_ssm_across_gdc',\n",
    "    '# CNV Gain':'nih_cnv_gain', '# CNV Loss':'nih_cnv_loss', 'Gene ID':'ensembl', '# Mutations':'nih_tot_mutations'}, axis=1)\n",
    "\n",
    "gdc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Feature Sparsity:\n",
      " Gene                                                     0\n",
      "Ensembl                                                  0\n",
      "Protein class                                            0\n",
      "Biological process                                    6858\n",
      "Molecular function                                    6651\n",
      "Disease involvement                                  10548\n",
      "Subcellular location                                  3666\n",
      "Tissue RNA - lung [NX]                                   0\n",
      "Single Cell Type RNA - Mucus-secreting cells [NX]        0\n",
      "dtype: int64\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\cbyle\\AppData\\Local\\Temp/ipykernel_24692/4004691990.py:23: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  hpa_features[\"Tissue RNA - lung [NX]\"] = (hpa_features[\"Tissue RNA - lung [NX]\"] - hpa_features[\"Tissue RNA - lung [NX]\"].mean()) / hpa_features[\"Tissue RNA - lung [NX]\"].std()\n",
      "C:\\Users\\cbyle\\AppData\\Local\\Temp/ipykernel_24692/4004691990.py:25: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  hpa_features[col] = (hpa_features[col] - hpa_features[col].mean()) / hpa_features[col].std()\n"
     ]
    }
   ],
   "source": [
    "hpa = pd.read_csv('../data/hpa_gene_features.tsv', sep='\\t').drop_duplicates(subset='Gene')\n",
    "\n",
    "identifiers = [\n",
    "    \"Gene\",\n",
    "    \"Ensembl\"\n",
    "]\n",
    "discrete_features = [\n",
    "    \"Protein class\",\n",
    "    \"Biological process\",\n",
    "    \"Molecular function\",\n",
    "    \"Disease involvement\",\n",
    "    \"Subcellular location\",\n",
    "]\n",
    "continuous_features = [\n",
    "    \"Tissue RNA - lung [NX]\",\n",
    "    \"Single Cell Type RNA - Mucus-secreting cells [NX]\"\n",
    "]\n",
    "\n",
    "hpa_features = hpa.iloc[:, hpa.columns.isin(identifiers+discrete_features+continuous_features)]\n",
    "\n",
    "\n",
    "print(\"Feature Sparsity:\\n\", hpa_features.isna().sum())\n",
    "hpa_features[\"Tissue RNA - lung [NX]\"] = (hpa_features[\"Tissue RNA - lung [NX]\"] - hpa_features[\"Tissue RNA - lung [NX]\"].mean()) / hpa_features[\"Tissue RNA - lung [NX]\"].std()\n",
    "col = \"Single Cell Type RNA - Mucus-secreting cells [NX]\"\n",
    "hpa_features[col] = (hpa_features[col] - hpa_features[col].mean()) / hpa_features[col].std()\n",
    "\n",
    "def explode(feature) :\n",
    "    return feature.apply(lambda x: x.replace(' ', '').split(','))\n",
    "\n",
    "hpa_clean = hpa.fillna('')\n",
    "for ft in discrete_features :\n",
    "    hpa_clean[ft] = explode(hpa_clean[ft])\n",
    "\n",
    "protein_class = hpa_clean[\"Protein class\"].explode().unique()\n",
    "biological_process = hpa_clean[\"Biological process\"].explode().unique()\n",
    "molecular_function = hpa_clean[\"Molecular function\"].explode().unique()\n",
    "disease_involvement = hpa_clean[\"Disease involvement\"].explode().unique()\n",
    "subcellular_location = hpa_clean[\"Subcellular location\"].explode().unique()\n",
    "GO_features = np.concatenate([protein_class, biological_process, molecular_function, disease_involvement, subcellular_location])\n",
    "\n",
    "RowFeatures = pd.DataFrame(data = 0,index = hpa_clean['Ensembl'],columns=GO_features)\n",
    "counter = 0\n",
    "\n",
    "hpa_clean\n",
    "\n",
    "for index, row in RowFeatures.iterrows() :\n",
    "    features = hpa_clean.iloc[counter][['Protein class', 'Biological process', 'Molecular function', 'Disease involvement', 'Subcellular location']].to_list()\n",
    "    flattened = [item for sublist in features for item in sublist if item]\n",
    "    for t in flattened :\n",
    "        row[t] = 1\n",
    "    counter +=1 \n",
    "\n",
    "n_comp = 100\n",
    "svd = TruncatedSVD(n_components = n_comp)\n",
    "svdModel = svd.fit(RowFeatures)\n",
    "visits_emb = svdModel.transform(RowFeatures)\n",
    "hpa = pd.DataFrame(data=visits_emb, index=RowFeatures.index)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import Disease Gene Network (DGN) data "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>EI_gda</th>\n",
       "      <th>Score_gda</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NAT2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ADA</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CDH2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CDKN2B-AS1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>HOTAIR</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2425</th>\n",
       "      <td>ZEB2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2426</th>\n",
       "      <td>CDC20</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2429</th>\n",
       "      <td>MFN2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2434</th>\n",
       "      <td>MVP</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2436</th>\n",
       "      <td>CDH1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.23</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>996 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Gene  EI_gda  Score_gda\n",
       "0           NAT2     1.0       0.02\n",
       "1            ADA     1.0       0.02\n",
       "2           CDH2     1.0       0.02\n",
       "3     CDKN2B-AS1     1.0       0.12\n",
       "6         HOTAIR     1.0       0.02\n",
       "...          ...     ...        ...\n",
       "2425        ZEB2     1.0       0.04\n",
       "2426       CDC20     1.0       0.03\n",
       "2429        MFN2     1.0       0.03\n",
       "2434         MVP     1.0       0.03\n",
       "2436        CDH1     1.0       0.23\n",
       "\n",
       "[996 rows x 3 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dgn = pd.read_csv(\"../data/gda_disease_summary_luad.csv\")\n",
    "\n",
    "score_threshold = 0.02\n",
    "ei_threshold = 0.7\n",
    "\n",
    "dgn = dgn[['Gene', 'EI_gda', 'Score_gda']]\n",
    "dgn = dgn.loc[dgn['Score_gda'] >= score_threshold]\n",
    "dgn = dgn.loc[dgn['EI_gda'] > ei_threshold]\n",
    "# dgn.rename({'Score_gda':'gda_score'}, axis=1, inplace=True)\n",
    "\n",
    "dgn"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "4410ce168b863aee9b4918bd897f8b9953e347697e7905a6c1ad714c78ed846f"
  },
  "kernelspec": {
   "display_name": "Python 3.8.6 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
